protocol-messages.go

```diff
@@ -0,0 +1,380 @@
+}
+
+//// Proposal
+type ProposalType uint8
+
+const (
+	invalid = iota
+	add
+	update
+	remove
+)
+
+type AddProposal struct {
+	ClientInitKey ClientInitKey
+}
+
+type UpdateProposal struct {
+	// this is a HPKEPublicKey (for this structure purposes it is opaque)
+	LeafKey []byte `tls:"head=2"`
+}
+
+type RemoveProposal struct {
+	Removed uint32
+}
+
+type Proposal struct {
+	Type   ProposalType
+	Add    *AddProposal
+	Update *UpdateProposal
+	Remove *RemoveProposal
+}
+
+func (p Proposal) MarshalTLS() ([]byte, error) {
+	s := &Stream{}
+	err := s.Write(p.Type)
+	if err != nil {
+		fmt.Errorf("mls.proposal: Marshal failed for ProposalType: %v", err)
+	}
+	syntax.Marshal(p.Type)
+	switch p.Type {
+	case add:
+		err = s.Write(p.Add)
+	case update:
+		err = s.Write(p.Update)
+	case remove:
+		err = s.Write(p.Remove)
+	default:
+		return nil, fmt.Errorf("mls.proposal: ProposalType type not allowed: %v", err)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("mls.proposal: Marshal failed: %v", err)
+	}
+	return s.Data(), nil
+}
+
+func (p *Proposal) UnmarshalTLS(data []byte) (int, error) {
+	s := &Stream{}
+	var ptype ProposalType
+	read, err := s.Read(data, &ptype)
+	if err != nil {
+		fmt.Errorf("mls.mlsproposal: Unmarshal failed for proposal type")
+		return 0, err
+	}
+	p.Type = ptype
+	switch p.Type {
+	case add:
+		var inner AddProposal
+		read, err := s.Read(data[read:], &inner)
+		if err != nil {
+			return 0, err
+		}
+		p.Add = &inner
+		return read, nil
+	case update:
+		var inner UpdateProposal
+		read, err := s.Read(data[read:], &inner)
+		if err != nil {
+			return 0, err
+		}
+		p.Update = &inner
+		return read, nil
+	case remove:
+		var inner RemoveProposal
+		read, err := s.Read(data[read:], &inner)
+		if err != nil {
+			return 0, err
+		}
+		p.Remove = &inner
+		return read, nil
+	default:
+		return 0, fmt.Errorf("mls.proposal: ProposalType type not allowed")
+	}
+}
+
+//// commit
+type ProposalId struct {
+	Sender uint32
+	Hash   []byte `tls:"head=1"`
+}
+
+type HPKECipherText struct {
+	KEMOutput  []byte `tls:"head=2"`
+	CipherText []byte `tls:"head=2"`
+}
+
+// struct {
+//		HPKEPublicKey public_key;
+//		HPKECiphertext encrypted_path_secret<0..2^16-1>;
+//  } DirectPathNode;
+//TODO: snk: Name doesn't match with the Spec
+type RatchetTreeNode struct {
+	// this is HPKEPublicKey
+	PublicKey           []byte           `tls:"head=2"`
+	EncryptedPathSecret []HPKECipherText `tls:"head=2"`
+}
+
+type DirectPath struct {
+	Nodes []RatchetTreeNode `tls:"head=2"`
+}
+
+type Commit struct {
+	Updates ProposalId
+	Removes ProposalId
+	Adds    ProposalId
+	Ignored ProposalId
+	Path    *DirectPath
+}
+
+/// MLSPlainText, MLSCipherText
+type epoch uint32
+
+const (
+	content_invalid = iota
+	content_application
+	content_proposal
+	content_commit
+)
+
+type ApplicationData struct {
+	Data []byte `tls:"head=4"`
+}
+
+type ProposalList struct {
+	Proposals []Proposal `tls:head=4`
+}
+
+type CommitData struct {
+	Proposals    ProposalList
```

Comment:
Note: Spec and `mlspp` differ on this.  `mlspp` omits the proposals list here, which avoids a lot of amiguity.

Reply:
i will get to this in the next round  as part of catching up on mlspp changes since the weekend